###                                       错误和异常处理



**show_error(*$message*, *$status_code*, *$heading = 'An Error Was Encountered'*)**



```php
/**
 *错误处理
    * 这个函数允许我们调用异常类 使用标准错误模板显示错误,模版在views/errors/error_general.php 
    * 这个函数将直接发送错误页面至浏览器并退出。
 * @param  string
 * @param  int
 * @param  string
 * @return void
 */
function show_error($message, $status_code = 500, $heading = 'An Error Was Encountered')
{
   $status_code = abs($status_code);
   if ($status_code < 100)
   {
      $exit_status = $status_code + 9; // 9 is EXIT__AUTO_MIN
      if ($exit_status > 125) // 125 is EXIT__AUTO_MAX
      {
         $exit_status = 1; // EXIT_ERROR
      }

      $status_code = 500;
   }
   else
   {
      $exit_status = 1; // EXIT_ERROR
   }

   $_error =& load_class('Exceptions', 'core');
   echo $_error->show_error($heading, $message, 'error_general', $status_code);
   exit($exit_status);
}
```

再定位到Exception类的show_error函数

```php
/**
 * 产生错误页面
 *
 *以一个错误消息作为输入(作为字符串或一个数组)并且使用指定的模板显示它
 * @param  string    $heading   Page heading
 * @param  string|string[]    $message   Error message
 * @param  string    $template  Template name
 * @param  int       $status_code   (default: 500)
 *
 * @return string Error page output
 */
public function show_error($heading, $message, $template = 'error_general', $status_code = 500)
{
   $templates_path = config_item('error_views_path');
   if (empty($templates_path))
   {
      $templates_path = VIEWPATH.'errors'.DIRECTORY_SEPARATOR;
   }

   if (is_cli())
   {
      $message = "\t".(is_array($message) ? implode("\n\t", $message) : $message);
      $template = 'cli'.DIRECTORY_SEPARATOR.$template;
   }
   else
   {
      set_status_header($status_code);
      $message = '<p>'.(is_array($message) ? implode('</p><p>', $message) : $message).'</p>';
      $template = 'html'.DIRECTORY_SEPARATOR.$template;
   }

   if (ob_get_level() > $this->ob_level + 1)
   {
      ob_end_flush();
   }
   ob_start();
   include($templates_path.$template.'.php');//这种方式非常巧妙
   $buffer = ob_get_contents();
   ob_end_clean();
   return $buffer;
}
```

**show_404($page = '', $log_error = TRUE)**

```php
/**
	 *404页面处理
	 *这个函数类似于show_error()函数。但是并不是用标准错误模板来显示404错误。
	 * @param	string
	 * @param	bool
	 * @return	void
	 */
	function show_404($page = '', $log_error = TRUE)
	{
		$_error =& load_class('Exceptions', 'core');
		$_error->show_404($page, $log_error);
		exit(4); // EXIT_UNKNOWN_FILE
	}
```



再定位到Exception类的show_404函数

```php
public function show_404($page = '', $log_error = TRUE)
	{
		if (is_cli())
		{
			$heading = 'Not Found';
			$message = 'The controller/method pair you requested was not found.';
		}
		else
		{
			$heading = '404 Page Not Found';
			$message = 'The page you requested was not found.';
		}

		// By default we log this, but allow a dev to skip it
		if ($log_error)
		{
			log_message('error', $heading.': '.$page);
		}

		echo $this->show_error($heading, $message, 'error_404', 404);
		exit(4); // EXIT_UNKNOWN_FILE
	}
```

可见，准备好数据之后，show_404函数最后还是调用show_error来展示错误。不过现在有个疑问，exit($exit_status),已知的是不同的数字代表了不同的退出状态，这个在constant.php文件中有详细定义。然而这样做的意义是什么？这些状态值被谁捕获？后期做什么响应和处理？据我所知如果PHP 的版本号大于等于 4.2.0，那么在 *status* 是整数的情况下，不会输出该状态值。这个还是留到后续的php源码实现剖析去解决吧。

**write_log($level, $msg)**

1. 运用了排他锁，LOCK_EX、LOCK_UN（ps：读的时候加上LOCK_SH能防止读到脏数据）
2. 如果是新文件，则赋权限644

```php
public function write_log($level, $msg)
	{
		if ($this->_enabled === FALSE)
		{
			return FALSE;
		}

		$level = strtoupper($level);

		if (( ! isset($this->_levels[$level]) OR ($this->_levels[$level] > $this->_threshold))
			&& ! isset($this->_threshold_array[$this->_levels[$level]]))
		{
			return FALSE;
		}

		$filepath = $this->_log_paath.'log-'.date('Y-m-d').'.'.$this->_file_ext;
		$message = '';

		if ( ! file_exists($filepath))
		{
			$newfile = TRUE;
			// Only add protection to php files
			if ($this->_file_ext === 'php')
			{
				$message .= "<?php defined('BASEPATH') OR exit('No direct script access allowed'); ?>\n\n";
			}
		}

		if ( ! $fp = @fopen($filepath, 'ab'))
		{
			return FALSE;
		}

		// Instantiating DateTime with microseconds appended to initial date is needed for proper support of this format
		if (strpos($this->_date_fmt, 'u') !== FALSE)
		{
			$microtime_full = microtime(TRUE);
			$microtime_short = sprintf("%06d", ($microtime_full - floor($microtime_full)) * 1000000);
			$date = new DateTime(date('Y-m-d H:i:s.'.$microtime_short, $microtime_full));
			$date = $date->format($this->_date_fmt);
		}
		else
		{
			$date = date($this->_date_fmt);
		}

		$message .= $level.' - '.$date.' --> '.$msg."\n";

		flock($fp, LOCK_EX);

		for ($written = 0, $length = strlen($message); $written < $length; $written += $result)
		{
			if (($result = fwrite($fp, substr($message, $written))) === FALSE)
			{
				break;
			}
		}

		flock($fp, LOCK_UN);
		fclose($fp);

		if (isset($newfile) && $newfile === TRUE)
		{
			chmod($filepath, $this->_file_permissions);
		}

		return is_int($result);
	}
```



**新技能get**



ob_get_level(PHP 4 >= 4.2.0, PHP 5)— 返回输出缓冲机制的嵌套级别

说明

int **ob_get_level** ( void )

返回输出缓冲机制的嵌套级别。返回值返回嵌套的输出缓冲处理程序的级别；或者是，如果输出缓冲区不起作用，返回零。



代码胜千言

```php
<?php

ob_start();

var_dump(ob_get_level());

ob_start();

var_dump(ob_get_level());

ob_start();

var_dump(ob_get_level());

ob_end_flush();

?>

```

运行效果：

int(1)
int(2)
int(3)

**用法1:错误捕获**

```php
//如果想要使用一个函数用于删除所有缓冲输出(为了只显示错误),需要清除所有级别的缓存
<?php 
function condor_error($errno, $errstr, $errfile, $errline) 
{ 
        $errors = array(E_USER_ERROR, E_ERROR, E_PARSE); 
        if (in_array($errno,$errors)) { 
                while (ob_get_level()) { 
                        ob_end_clean(); 
                } 
                echo "<B>FATAL</B> [".$errno."] ".$errstr."<br> "; 
                echo "Fatal error at line ".$errline." of file ".$errfile; 
                echo ", PHP ".PHP_VERSION." (".PHP_OS.")<br> "; 
                echo "Aborting...<br><br> "; 
                exit; 
        } 
} 

set_error_handler("condor_error"); 
?>
```

**用法2：异常处理**

```php
function test()
{
    $level = ob_get_level();
    try
    {
        ob_start();
        echo 'some msg1';
        ob_start();
        echo 'some msg2';
        throw new Exception('happen Exception');
    }

    catch (Exception $e)
    {
        //清理msg1和msg2
        //while (ob_get_level() > $level)
        //{
            //ob_end_clean();
        //}

        var_dump($e->getMessage());
    }
}
test();
```